docs/rfcs/021-raii-runtime.md

```diff
@@ -0,0 +1,209 @@
+
+- It needs to be clarified how to know that tenant is being dropped.
+- `Active` state has an option that's not needed. If we do not want to run background tasks, we may wish to have an alternative to the `Active` state without running background jobs.
+- Actually, we can change the state from `Paused` to `Active` _and vice versa_! Engineers must remember that when writing the code, the only reason to get to know that is to read the code. Only some of the people in the team know about that.
+
+Proposed states:
+
+1. `Infant` - just created, no activity yet.
+    - Do we need this state?
+2. `Loading` - currently loads its data from the disk to the memory.
+3. `Broken` - cannot load the tenant; unrecoverable error happened. It is the final state.
+4. `Downloading` - downloading the data files from object storage.
+    - Is it better to change to `Attaching`?
+5. `Active` - fully operational, background tasks are running.
+6. `ShuttingDown` - tenant is being shut down; no new closures could be run.
+    - What should happen to operations running on the tenant when we change the state to this one?
+    - Is it better to change to `Terminating`?
+7. `Shutdown` - tenant is shut down, background operations are finished, and ready to recycle. It is the final state.
+    - Is it better to change to `Terminated`?
+
+Transitions:
+
+1. `Infant` -> `Loading`, `Downloading`.
+2. `Loading` -> `Active`, `Broken`.
+3. `Broken` is the final state.
+4. `Downloading` -> `Active`, `Broken`.
+5. `Active` -> `ShuttingDown`.
+6. `ShuttingDown` -> `Shutdown`.
+7. `Shutdown` is the final state.
+
+The transition graph has no cycles, which is a good property.
+
+Generally, we have 3 subsets of states:
+
+1. Paused - tries to load itself: `Loading`, `Downloading`.
+2. Alive - can operate: `Infant`, `Active`.
+3. Dead - cannot launch any operations, possibly already dropped: `Broken`, `ShuttingDown`, `Shutdown`.
+
+## `TenantAccessor`
+
+The first part of the chapter will be about `TenantAccessor` specifically. Later, we'll discuss `TimelineAccessor`.
+
+### Top-level interface
+
+We need a proper tree-like structure with a bare minimum of global states to use RAII, unlike carefully using `shutdown_tasks` as we do now.
+
+The idea is to have the following tenant API:
+
+```rust
+/// Gets the alive tenant from the in-memory data, then applies the specified
+/// function. Returns error if there is no tenant in memory or the tenant is not
+/// ready to launch the closure.
+pub fn with_tenant<F, O>(tenant_id: TenantId, func: F) -> anyhow::Result<O>
+where
+    F: FnOnce(TenantSyncRef) -> O {}
+
+/// Gets the alive tenant from the in-memory data, then applies the specified
+/// asynchronous function. Returns error if there is no tenant in memory or
+/// tenant is not ready to launch the closure.
+pub async fn with_tenant_async<F, T, O>(tenant_id: TenantId, func: F) -> anyhow::Result<O>
+where
+    F: FnOnce(TenantAsyncRef) -> T,
+    T: Future<Output = O> {}
+```
+
```

Comment:
I'm not a fan of this closure pattern, I think a guard object, similar to e.g. Mutex and MutexGuard, would be better. I don't think it changes the big picture though, AFAICS either approach would work more or less the same.

So instead of `with_tenant(...)`, I'm imagining that the code that uses this would look something like this:
```
{
    let tenant_guard = get_active_tenant(tenant_id);
    tenant_guard.do_stuff();
}
```

Reply:
IMO it's more error-prone to have a guard. We'll have to have another scope like in your example or use `drop` manually. The first one is the same as having `with_tenant` even comparing the code tabulation, and the second requires `drop`, which could be forgotten when writing and reviewing the code, making the scope with this guard bigger than expected and probably accidentally including some waiting and things like that, that could break the code.